Refactored GlusterFS support

hlub opened the pull request:
Usage of GlusterFS states did not change except the state 'created' was deprecated in favor of more descriptive name 'volume_present'.
Improved separation of functionality in the execution module from the states. Module functions now return booleans instead of varying
strings. Parsing those strings in the sates were very error-prone. Improved checking and verbosity in states. Calls to gluster are now logged more verbosely,
while a failing call no longer raises an immediate exception. Cleaned code a lot, removed some bugs and other odd behaviour. It could be beneficial to return
dictionaries with boolean status and gluster output string from some of the module functions but please don't do error checking based on the messages!
Some other functional changes:
- glusterfs.info() can now be called without volume name. It always returns a dictionary. Use this function instead of status() in sates.
- glusterfs.list_peers() function now attempts to list the longest possible form of the hostnames. GlusterFS may list several hostnames from which one
  is a default. An alternative hosts is used instead if it is prefixed by the default. (e.g. gluster1 is replaced with gluster1.example.com).
- Functions glusterfs.delete() and glusterfs.create() were renamed to delete_volume() and create_volume().

hlub committed:
Refactored GlusterFS support

Usage of GlusterFS states did not change except the state 'created' was deprecated in favor of more descriptive name 'volume_present'.
Improved separation of functionality in the execution module from the states. Module functions now return booleans instead of varying
strings. Parsing those strings in the sates were very error-prone. Improved checking and verbosity in states. Calls to gluster are now logged more verbosely,
while a failing call no longer raises an immediate exception. Cleaned code a lot, removed some bugs and other odd behaviour. It could be beneficial to return
dictionaries with boolean status and gluster output string from some of the module functions but please don't do error checking based on the messages!
Some other functional changes:
- glusterfs.info() can now be called without volume name. It always returns a dictionary. Use this function instead of status() in sates.
- glusterfs.list_peers() function now attempts to list the longest possible form of the hostnames. GlusterFS may list several hostnames from which one
is a default. An alternative hosts is used instead if it is prefixed by the default. (e.g. gluster1 is replaced with gluster1.example.com).
- Functions glusterfs.delete() and glusterfs.create() were renamed to delete_volume() and create_volume().

gtmanfred commented:
I would say that the create and delete should be aliases, at least until nitrogen as well, so as not to break compatibility for anyone using the modules.

and you have a couple pylint errors https://jenkins.saltstack.com/job/salt-pr-lint-n/13499/console

And some errors in the glusterfs tests https://jenkins.saltstack.com/job/salt-pr-rs-cent7-n/12519/

thanks!
Daniel

joejulian commented on salt/modules/glusterfs.py:
```diff
@@ -43,34 +43,46 @@ def _get_minor_version():
-    return __salt__['cmd.run'](
-        'gluster --mode=script', stdin="{0}\n".format(cmd))
+    return int(xml_data.find('opRet').text) == 0
 
```
There are a few gluster commands that don't export xml. The _gluster command needs to remain to handle those exceptions.

joejulian commented on salt/modules/glusterfs.py:
```diff
@@ -43,34 +43,46 @@ def _get_minor_version():
+            log.info('Gluster call "{0}" succeeded'.format(cmd))
+    else:
+        log.error('Failed gluster call: {0}: {1}'.format(cmd, root.find('opErrstr').text))
     return root
```
Modules can raise errors. This is handled gracefully elsewhere.

hlub replied:
Yes, they can raise errors and the code still raises some (salt invocation errors). A good question is, which types of errors should do that? Just every error or only a subset like bad invocations. I've no answers or even strong opinions about this. In addition to the internal use in states, module functions are directly callable by users and I was thinking that nice return values would be appreciated there. Although, I don't know how Salt reports exceptions in that case. I found the current behaviour of immediately raising an exception for a failing Gluster call problematic as it produced some bugs. For instance, a failing status() call raised an unhandled exception while the calling code was supposed to do things when an error occurs. Of course, this could've been fixed by improving exception handling. My modifications may loose some verbosity (excluding logs) but this could be fixed by returning dictionaries with both boolean error and output string. At least, I hope that after my modifications the code is easily improvable despite of which decisions are made on this specific question.

joejulian commented on salt/modules/glusterfs.py:
```diff
@@ -43,34 +43,46 @@ def _get_minor_version():
             'gluster --xml --mode=script', stdin="{0}\n".format(cmd)
         ).replace("\n", ""))
-    if int(root.find('opRet').text) != 0:
-        raise CommandExecutionError(root.find('opErrstr').text)
```
Notice, in my most recent PR, that I'm no longer using the string output for determining results.

joejulian commented on salt/modules/glusterfs.py:
```diff
@@ -117,11 +133,20 @@ def list_peers():
+                    hostname = alternative.text
+        result.append(hostname)
+
+    return result
```
Similar idea, but instead I'm outputting a dict with the server's hostname being the key for a list of alternate hostnames. This allows future matches against any valid means of identifying said host.

hlub replied:
I think a simple list of consistent peer names would be useful especially in Jinja templates and other simple cases. It could be nice to add another function like peer_status(), which would return a richer set of info as the corresponding Gluster call does.

joejulian replied:
I thought about that, and chose a dict with the hostname as the key for the aliases because any code that used the list of hostnames **as** a list will still work. An iterator is still going to walk through the keys, which will be the list of hostnames.

joejulian commented on salt/states/glusterfs.py:
```diff
@@ -48,57 +49,59 @@ def peered(name):
+        ret['result'] = True
+        ret['comment'] = 'Host {0} already peered'.format(name)
+        return ret
 
```
Removing the 'test' option would have broken the unit tests. I have also removed it as I have reworked the unit tests to allow actual logic testing.

joejulian commented on salt/states/glusterfs.py:
```diff
@@ -48,57 +49,59 @@ def peered(name):
 
 
-def created(name, bricks, stripe=False, replica=False, device_vg=False,
+def volume_present(name, bricks, stripe=False, replica=False, device_vg=False,
```
I'm in the process of completely reworking this to be much more intelligent, allowing you to change the volume definition by just passing it a changed state tree, eg add/remove/replace bricks/servers/replica levels, etc.

joejulian commented:
Great work. Nice to see were were both thinking along the same lines. Hit me up on IRC in #gluster or #salt if you'd like to chat about the work I'm doing on improving the gluster components for salt to see if we can collaborate.

hlub committed:
Update GlusterFS unit tests to cover the refactored code

Fixed and updated failing tests and cleaned some testing code (especially in states.glusterfs_test). Added assertions to test that some functions are not called when things are already in the desired state. Preferred changing behaviour of the mocked functions while proceeding instead of listing all of the return values first.

cachedout commented:
Thanks for the discussion here. @hlub I see you added some commits after the comments from @joejulian 

Could one or both of you let me know when this is satisfactory to you both and I'll do a final review and hopefully get this merged? Thanks.

joejulian commented:
This needs to be rebased against 2015.8 and I'd prefer that it happen after #31287 is merged.

cachedout commented:
@joejulian I do have #31287 in.

@hlub If you could re-submit this against 2015.8 and also fix the lint errors, we can take it from there. Thanks.

hlub committed:
Minor GlusterFS code refining

Fixed the lint errors and replaced the deprecated created() with the volume_present() in the tests. Added deprecated delete() and create() as it was suggested in the PR conversation.

hlub commented:
I cherry-picked the changes under 2015.8, and did my best to take the changes of @joejulian into account. The re-submitted PR is #31395. Hopefully this is enough. I can't put much more time on this right now. I've written a state called glusterfs.options_reconfigured for volume configuration management but planned to submit it in a separate PR.

cachedout commented:
@hlub OK, I will close this one in favor of #31395 then. Thanks.
